app/src/main/java/com/kickstarter/viewmodels/MessagesViewModel.java

```diff
@@ -75,129 +103,211 @@ public ViewModel(final @NonNull Environment environment) {
-        backing.map(ObjectUtils::isNull)
-      )
-        .distinctUntilChanged()
+      participant
+        .map(User::name)
         .compose(bindToLifecycle())
-        .subscribe(this.backingInfoViewHidden::onNext);
+        .subscribe(this.participantNameTextViewText::onNext);
+
+      this.messageEditTextHint = this.participantNameTextViewText;
 
       Observable.combineLatest(
-        backing,
-        messageThread.map(MessageThread::project),
-        Pair::create
+        backingOrThread,
+        project,
+        participant,
+        this.currentUser.observable(),
+        MessagesData::new
       )
+        .switchMap(data -> backingAndProjectFromData(data, this.client))
         .compose(bindToLifecycle())
         .subscribe(this.backingAndProject::onNext);
 
-      messageThread
-        .map(thread -> thread.project().creator().name())
+      this.backingAndProject
+        .map(bp -> bp.first == null)
         .compose(bindToLifecycle())
-        .subscribe(this.participantNameTextViewText::onNext);
+        .subscribe(this.backingInfoViewIsGone::onNext);
 
-      this.messageEditTextHint = this.participantNameTextViewText;
-
-      messageThreadEnvelope
-        .map(MessageThreadEnvelope::messages)
+      koalaContext
+        .map(c -> c.equals(KoalaContext.Message.BACKER_MODAL))
         .compose(bindToLifecycle())
-        .subscribe(this.messages::onNext);
+        .subscribe(this.viewPledgeButtonIsGone::onNext);
+
+      this.backButtonIsGone = this.viewPledgeButtonIsGone.map(BooleanUtils::negate);
+      this.closeButtonIsGone = this.backButtonIsGone.map(BooleanUtils::negate);
+      this.goBack = this.backOrCloseButtonClicked;
 
       messageNotification
         .compose(errors())
         .map(ErrorEnvelope::fromThrowable)
         .map(ErrorEnvelope::errorMessage)
         .subscribe(this.showMessageErrorToast::onNext);
 
-      messageThread
-        .map(thread -> thread.project().name())
+      project
+        .map(Project::name)
         .compose(bindToLifecycle())
         .subscribe(this.projectNameTextViewText::onNext);
+
+      project
+        .compose(takeWhen(this.viewPledgeButtonClicked))
+        .compose(bindToLifecycle())
+        .subscribe(this.startViewPledgeActivity::onNext);
+
+      project
+        .take(1)
+        .compose(bindToLifecycle())
+        .subscribe(this.koala::trackViewedMessageThread);
+
+      Observable.combineLatest(project, koalaContext, Pair::create)
+        .compose(takeWhen(messageSent))
+        .compose(bindToLifecycle())
+        .subscribe(pc -> this.koala.trackSentMessage(pc.first, pc.second));
     }
 
+    private static @NonNull Observable<Pair<Backing, Project>> backingAndProjectFromData(final @NonNull MessagesData data,
+      final @NonNull ApiClientType client) {
+
+      return data.getBackingOrThread().either(
+        backing -> Observable.just(Pair.create(backing, data.getProject())),
```

Comment:
FYI this `getSomething()` syntax is because `MessagesData` is a kotlin data class 🎈